File: domain-services/finance/savvy-service/src/main/java/com/ayman/distributed/savvy/services/investment/InvestmentService.java
package com.ayman.distributed.savvy.services.investment;

import com.ayman.distributed.savvy.model.dto.InvestmentDTO;
import com.ayman.distributed.savvy.model.entity.Investment;

import java.util.List;

public interface InvestmentService {
    Investment createInvestment(InvestmentDTO investmentDTO);
    List<Investment> getAllInvestments();
    Investment getInvestmentById(Long id);
    Investment updateInvestment(InvestmentDTO investmentDTO, Long id);
    void deleteInvestment(Long id);
}

File: client-applications/savvy-web/src/app/services/investment.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Investment } from '../models/dashboard';

@Injectable({
  providedIn: 'root',
})
export class InvestmentService {
  private http = inject(HttpClient);
  private apiUrl = '/api/investment';

  createInvestment(
    investmentData: Partial<Investment>,
  ): Observable<Investment> {
    return this.http.post<Investment>(this.apiUrl, investmentData);
  }

  getAllInvestments(): Observable<Investment[]> {
    return this.http.get<Investment[]>(`${this.apiUrl}/all`);
  }

  getInvestmentById(id: number): Observable<Investment> {
    return this.http.get<Investment>(`${this.apiUrl}/${id}`);
  }

  updateInvestment(
    id: number,
    investmentData: Partial<Investment>,
  ): Observable<Investment> {
    return this.http.put<Investment>(`${this.apiUrl}/${id}`, investmentData);
  }

  deleteInvestment(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
